atoma-p2p/src/metrics.rs

```diff
@@ -0,0 +1,123 @@
+    // pub uptime: u64,
+    /// The CPU usage of the node
+    pub cpu_usage: f32,
+    /// The amount of RAM used
+    pub ram_used: u64,
+    /// The total amount of RAM in the system
+    pub ram_total: u64,
+    /// The amount of RAM used in swap
+    pub ram_swap_used: u64,
+    /// The total amount of swap memory in the system
+    pub ram_swap_total: u64,
+    /// The number of CPUs in the system
+    pub num_cpus: u32,
+    /// The total number of bytes received from the network
+    pub network_rx: u64,
+    /// The total number of bytes transmitted to the network
+    pub network_tx: u64,
+    /// The number of GPUs in the system
+    pub num_gpus: u32,
+    /// The usage metrics for each GPU
+    pub gpus: Vec<GpuMetrics>,
+}
+
+/// Structure to store the usage metrics for each GPU
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct GpuMetrics {
+    /// The utilization of the GPU
+    pub utilization: u32,
+    /// The amount of memory used by the GPU
+    pub memory_used: u64,
+    /// The total amount of memory on the GPU
+    pub memory_total: u64,
+    /// The amount of free memory on the GPU
+    pub memory_free: u64,
+    /// The percentage of time the GPU was reading or writing
+    pub percentage_time_read_write: u32,
+    /// The percentage of time the GPU was executing
+    pub percentage_time_gpu_execution: u32,
+    /// The temperature of the GPU
+    pub temperature: u32,
+    /// The power usage of the GPU
+    pub power_usage: u32,
+}
+
+/// Returns the usage metrics for the node
+#[instrument(level = "info", target = "metrics")]
+pub fn compute_usage_metrics(mut sys: System) -> Result<NodeMetrics, NodeMetricsError> {
+    let nvml = Nvml::init()?;
+
+    let device_count = nvml.device_count()?;
+    let mut gpus = Vec::new();
+    for i in 0..device_count {
+        let device = nvml.device_by_index(i)?;
+        let Utilization { gpu, memory } = device.utilization_rates()?;
+        let MemoryInfo { used, total, free } = device.memory_info()?;
+        let temperature = device.temperature(TemperatureSensor::Gpu)?;
+        let power_usage = device.power_usage()?;
+        gpus.push(GpuMetrics {
+            utilization: gpu,
+            memory_used: used,
+            memory_total: total,
+            memory_free: free,
+            percentage_time_read_write: memory,
+            percentage_time_gpu_execution: gpu,
+            temperature,
+            power_usage,
+        });
+    }
+
+    let cpu_usage = sys.global_cpu_usage();
+    let ram_used = sys.used_memory();
+    let ram_total = sys.total_memory();
+    let ram_swap_used = sys.used_swap();
+    let ram_swap_total = sys.total_swap();
+    let num_cpus = sys.cpus().len();
+    let networks = Networks::new_with_refreshed_list();
+    let mut network_rx = 0;
+    let mut network_tx = 0;
+    for (_interface, data) in &networks {
+        network_rx += data.received();
+        network_tx += data.transmitted();
+    }
+
+    sys.refresh_all();
```

Comment:
Perhaps we should move this to line 86 before we collect the metrics to avoid staleness

Reply:
Definitely agree, great catch !